cloud-aws-cloudformation/src/main/resources/templates/aws-cf-stack.ftl

```diff
@@ -315,6 +315,12 @@
           "TargetType" : "instance",
           "HealthCheckPort" : "${listener.targetGroup.healthCheckPort}",
           "HealthCheckIntervalSeconds" : 10,
+          <#if loadBalancer.awsScheme == "internal">
```

Comment:
What's the purpose of this change? Is it just to get around not being able to loopback traffic? Keep in mind that with this, we lose all context for where the traffic originated. I know this is the internal load balancer, so it might not matter as much, but we're losing the ability to filter/set security rules for traffic based on where it originated, because when the traffic gets to the instance it looks like it originated at the load balancer.

Reply:
Actually, the security rules are a nonissue for internal load-balancer. We must allow all the traffic between instances in the same subnet. The IP for audit purposes is a legitimate concern. The only use case for the internal load-balancer is Hue/Yarn to Oozie - service to service communication. Which is more of a log to troubleshoot than to do a serious audit about it.

Reply:
The internal load balancer is also used for all data lakes that have private subnets, to proxy to Knox. All data lakes have a load balancer. The load balancer is of type internal if the data lake uses private subnets, and internet-facing if the data lake uses public subnets. Basically:
* We're registering the LB FQDN in a public AWS Route 53 zone, because that's what we do for all FQDNs, public and private.
* We're using an NLB, so we don't have any way to add a security policy to the LB, so we don't have a way to block traffic at the load balancer level and need to rely on the EC2 instance security policy.
* With this change, we're making it look like all traffic is coming from the NLB. Which means that as long as the security policy is set up to accept traffic from the load balancer, which is must be, then it will accept all traffic sent to the load balancer, regardless of that traffic's source.

What I think all that means is if someone can find the LB's FQDN, which again is technically public, they can make calls through it regardless of the EC2 instance security policy, because the instance can't tell where the traffic is coming from with this change.

From the testing I did around this previously: "I think I just confirmed the security group thing. I used the existing load balancer config, and modified my security group to only accept HTTPS traffic from the VPC CIDR. I verified I could make an API call through the LB successfully from one of the cluster nodes, but couldn't from my laptop. I then changed the target group to use IP addresses instead of instances ids, and was able to make an API call from my laptop"

What I don't entirely remember is whether or not I was using an internal or an internet-facing NLB there. But based on the JIRA (CB-11925) which was not about the public load balancer, I'm about 90% sure I was testing on an internal NLB. I really think this configuration is going to allow all traffic to flow through the NLB with no way to filter it. That might not matter too much for Hue, Yarn, and Oozie, but it matters a lot for Knox.

Reply:
The concept of an internal load-balancer is it is not publicly routable. Only the internet-facing load-balancers are publicly routable. We might have registered it publicly but that does not mean anyone can access it. The internal load-balancers have private IP. Only those who have access to that IP can access any service which is behind the IP. Those who have that access to the internal load-balancer already have access to the service endpoint also. That is why I deem it a nonissue.

> What I don't entirely remember is whether or not I was using an internal or an internet-facing NLB there. But based on the JIRA (CB-11925) which was not about the public load balancer, I'm about 90% sure I was testing on an internal NLB. I really think this configuration is going to allow all traffic to flow through the NLB with no way to filter it. That might not matter too much for Hue, Yarn, and Oozie, but it matters a lot for Knox.

I am 100% sure it is internal load-balancer. You are accessing a load-balancer port 443 which has private IP. Which is impossible unless there is a VPN connection or AWS transit gateway. So the trust is already established via a route with the help of firewall rules at the VPC layer. Now you want to specifically restrict a particular instance's port 443 to restrict inbound traffic from the load-balancer when the request origin is outside of VPC CIDR? What is the use of load-balancer here then?

> I then changed the target group to use IP addresses instead of instances ids, and was able to make an API call from my laptop

Though it does not matter, I would like to point out that, I am not using IP addresses instead of instance ids. All I am doing is not preserving the client IP at the load-balancer layer.

Reply:
Resolved this by moving Hue away from the Oozie hostgroup.